### PR TITLE
[Parse] Improve error handling in parseList

### PR DESCRIPTION
**Pull request by rintaro**
- Skip ahead if seeing any error while parsing list elements.<br>
  In most cases, we cannot expect that the following tokens could construct a valid element (with missing separator). Skip them, instead of trying to parse them as a next element.
- Bailout if seeing `tok::eof` or token that can never start a element, after parsing an element.<br>
  This silences superfluous `expected ',' separator` error, or misleading `expected declaration` error. What we should emit is `expected ')' in expression list` or `expected '}' in struct`.
- Don't emit missing closing token error if we've already emitted any error while parsing elements.<br>
  **[RFC] about this change**: This might be too silence, but some of other places like this don't emit error for missing closing token.

### CODE REVIEW

File: validation-test/compiler_crashers_fixed/00058-get-self-type-for-container.swift [hunk 1]
```diff
@@ -6,6 +6,6 @@
 // See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 
 // RUN: %target-swift-frontend %s -parse -verify
```
Comment (slavapestov) [inline validation-test/compiler_crashers_fixed/00058-get-self-type-for-container.swift]: compiler_crashers tests should not use -verify -- just remove the expected-errors here and the -verify flag

Comment (rintaro) [inline validation-test/compiler_crashers_fixed/00058-get-self-type-for-container.swift]: done.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by DougGregor** [comment]
This LGTM, and I'm fine with omitting the closing-token error when there are other errors: we're basically trying to resync to something sane at that point anyway.

**Comment by rintaro** [comment]
@DougGregor Thank you!
I'm going to merge this after re-running CI.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Pull request merged**